src/plugins/nest.js

```diff
@@ -0,0 +1,576 @@
+          if (pipes.length > 0) {
+            if (pipes[0].constructor && pipes[0].constructor.name) {
+              opName = `${pipes[0].constructor.name}.pipeFn`
+            }
+          }
+          const scope = tracer.scope()
+          const childOf = scope.active()
+          const span = tracer.startSpan(opName, {
+            childOf,
+            tags: {
+              'component': 'nest'
+            }
+          })
+          if (paramsOptions && paramsOptions[0]) {
+            const pipes = []
+            const pipeOptions = paramsOptions[0].pipes
+            pipeOptions.forEach((param) => {
+              if (param.constructor && param.constructor.name) {
+                pipes.push(param.constructor.name)
+              }
+            })
+            if (pipes.length > 0) {
+              span.setTag('nest.pipes', pipes)
+            }
+          }
+
+          return scope.activate(span, () => {
+            try {
+              return pipesFn.apply(this, [args, req, res, next])
+            } catch (e) {
+              throw addError(span, e)
+            } finally {
+              span.finish()
+            }
+          })
+        }
+      }
+      return wrappedPipesFn(createPipesFn.apply(this, arguments))
+    }
+  }
+}
+
+function createGuardsTrace (tracer, args, guards, instance, callback, fn) {
+  let opName = 'nest.guard.canActivate'
+  const request = args.length > 1 ? args[0] : args
+  const scope = tracer.scope()
+  const childOf = scope.active()
+  const span = tracer.startSpan(opName, {
+    childOf,
+    tags: {
+      'component': 'nest',
+      'request.method': request.method,
+      'request.url': request.originalUrl,
+      'request.route.path': request.route.path
+    }
+  })
+
+  const guardNames = []
+  guards.forEach(guardName => {
+    guardNames.push(guardName.constructor.name)
+  })
+  if (guardNames.length > 0) {
+    if (guardNames[0].constructor && guardNames[0].constructor.name) {
+      opName = `${guardNames[0]}.tryActivate`
+    }
+    span.setTag('nest.guards', guardNames)
+  }
+  if (instance.constructor && instance.constructor.name) {
+    opName = `${opName}.${instance.constructor.name}`
+    span.setTag('nest.controller.instance', instance.constructor.name)
+  }
+  if (callback.name) {
+    opName = `${opName}(${callback.name})`
+    span.setTag('nest.callback', callback.name)
+  }
+
+  span.setOperationName(opName)
+
+  return scope.activate(span, () => {
+    try {
+      return fn.apply(this, args)
+    } catch (e) {
+      throw addError(span, e)
+    } finally {
+      span.finish()
+    }
+  })
+}
+
+function addError (span, error) {
+  span.setTag('error', 'true')
+  span.log({
+    'error.type': error.name,
+    'error.msg': error.message,
+    'error.stack': error.stack
+  })
+  return error
+}
+
+function patchNestFactory (versions, additionalDependencies) {
+  return {
+    name: '@nestjs/core',
+    versions: versions,
+    additionalDependencies: additionalDependencies,
```

Comment:
Ideally we'd not need to change the patch signature like this for test purposes.  Why are we not using the existing `test/plugins/externals.json` and modifying `withVersions` as necessary?

Reply:
I didn't know it existed. The additionalDependencies feature was added by @justindsmith so he might know why externals.json was not enough. I took a quick look at externals.json and it seems that it allows us to install additional packages per instrumented package version. For example, the following snippet means that we want to install `@nestjs/websockets` package for `nest` plugin when nest version is one of '2.1.1', '~3.*' and '^4.*'.

```json   
    nest: [ 
       {
          "name": "@nestjs/websockets",
          "versions": ["2.1.1", "~3.*","^4.*"]
       }
   ]
```

Is that correct? If so, is there a way to specify exactly which version of nestjs/websockets should be installed per version of the instrumented plugin? This is required because we need different versions of such dependencies for different @nestjs/core versions. If this is possible then we don't need the additionalDependencies feature. 

If it is not possible with externals.json, I can may be re-work the additionalDependencies feature as an extension of externals.json.

--

Personally, right now, I'm not sure if externals.json is better than the additionalDependencies mainly because I did not discover it yet while trying to solve a complex dependency resolution issue. Perhaps mentioning such deps close to plugin patch definition would be better for discoverability and debugging in future.

Reply:
@justindsmith did you try to use externals.json and if so, do you remember why you opted for additionalDependencies pattern instead of using externals.json?

Reply:
> so he might know why externals.json was not enough
He did not know that existed!

I was just trying to get a "okay, this highlights where the problem is (it's a version problem), and offers a strawman solution, but I don't know if this is the right architectural pattern to use in this code base". So if the externals.json can handle the different dependency versions depending on test then we should definitely go that direction.

One reason I DON'T like the additionalDependencies where I put them (in `src/plugins/nest.js` is that they are really related to the test artifacts and not the core plugin work, so it's not good to pollute the core library with that additional complexity.

Reply:
First look tell me it does not but I might be wrong. I'll dig deeper tomorrow and see if I can add support to externals.json if needed.

Reply:
As far as I could tell, externals.json does not solve our problem. I added a similar dependencies.json to make it work for us. The feature is minimal and self-contained in tests (does not touch normal operations).